Heap pollution: Deserializing into @JsonbCreator with Set<Thing> creates a Set<HashMap> instead

I am trying to deserialize `{"persons": [{"name": "name1"}]}` into a class [`Persons`](https://github.com/oliviercailloux/2018-jsonb-serial/blob/master/src/main/java/io/github/oliviercailloux/y2018/jsonb_serial/Persons.java). The class declares a JsonbCreator as follows:

	@JsonbCreator
	public static Persons wrap(@JsonbProperty("persons") Set<Person> persons)

Jsonb knows how to create a [`Person`](https://github.com/oliviercailloux/2018-jsonb-serial/blob/master/src/main/java/io/github/oliviercailloux/y2018/jsonb_serial/Person.java) from the string `{"name": "name1"}` (I have checked with another [unit test](https://github.com/oliviercailloux/2018-jsonb-serial/blob/master/src/test/java/io/github/oliviercailloux/y2018/jsonb_serial/PersonsTest.java)).

What jsonb actually puts in the variable `persons`, [however](https://en.wikipedia.org/wiki/Heap_pollution), is a `Set<HashMap>`, as determined when trying to access the first person in `persons`, which throws an exception:

	javax.json.bind.JsonbException: Exception occurred during call to JSONB creator on class: class io.github.oliviercailloux.y2018.jsonb_serial.Persons.
	Caused by: java.lang.reflect.InvocationTargetException
	Caused by: java.lang.ClassCastException: java.util.HashMap cannot be cast to io.github.oliviercailloux.y2018.jsonb_serial.Person

The full code is [here](https://github.com/oliviercailloux/2018-jsonb-serial).

--- a/src/main/java/org/eclipse/yasson/internal/model/CreatorModel.java
+++ b/src/main/java/org/eclipse/yasson/internal/model/CreatorModel.java
@@ -43,7 +43,7 @@ public class CreatorModel {
      */
     public CreatorModel(String name, Parameter parameter, JsonbContext context) {
         this.name = name;
-        this.type = parameter.getType();
+        this.type = parameter.getParameterizedType();
 
         AnnotationIntrospector annotationIntrospector = context.getAnnotationIntrospector();
 

--- a/src/test/java/org/eclipse/yasson/customization/JsonbCreatorTest.java
+++ b/src/test/java/org/eclipse/yasson/customization/JsonbCreatorTest.java
@@ -26,6 +26,7 @@ import javax.json.bind.annotation.JsonbNumberFormat;
 import javax.json.bind.annotation.JsonbProperty;
 import java.math.BigDecimal;
 import java.time.LocalDate;
+import java.util.Set;
 
 import static org.junit.Assert.*;
 
@@ -141,6 +142,45 @@ public class JsonbCreatorTest {
         ParameterNameTester result = JsonbBuilder.create().fromJson(json, ParameterNameTester.class);
         Assert.assertEquals("someText", result.name);
         Assert.assertNull(result.secondParam);
+    }
+
+    @Test
+    public void testGenericCreatorParameter() throws Exception {
+        final String json = "{\"persons\": [{\"name\": \"name1\"}]}";
+        Jsonb jsonb = JsonbBuilder.create();
+        Persons persons = jsonb.fromJson(json, Persons.class);
+        Assert.assertEquals(1, persons.hiddenPersons.size());
+        Assert.assertEquals("name1", persons.hiddenPersons.iterator().next().getName());
+    }
+
+    public static final class Persons {
+
+        Set<Person> hiddenPersons;
+
+        private Persons(Set<Person> persons) {
+            this.hiddenPersons = persons;
+        }
+
+        @JsonbCreator
+        public static Persons wrap(@JsonbProperty("persons") Set<Person> persons) {
+            return new Persons(persons);
+        }
+
+        public Set<Person> getPersons() {
+            return null;
+        }
+    }
+
+    public static final class Person {
+        private String name;
+
+        public String getName() {
+            return name;
+        }
+
+        public void setName(String name) {
+            this.name = name;
+        }
     }
 
 

